### ISSUE TITLE
Clock on Nia is losing time

### ISSUE THREAD

**Issue by WS64** [koreader/koreader#7092]
* KOReader version: 2020-12
* Device: Kobo Nia

#### Issue
Kobo Nia seems to lose contact with the internal clock.
When I start Koreader the clock is correct, the longer I use it the more it falls behind. Rebooting/going back to nickel and stuff like that always brings back the correct time. But staying inside Koreader the clock is too slow. It seems to depend on what I do, at some times it does not lose much at all, in some times it loses half of the real time, e.g. when I start it at 9am and read for 10 minutes the Koreader clock tells me it's 9:05 instead of 9:10.
This behavior is confirmed by other users in a German forum, and only for Kobo Nia.

**Comment by Frenzie** [comment]
Duplicate of https://github.com/koreader/koreader/issues/6968.

When in Nickel, is wifi enabled? Because that automatically does a time sync. We have a synchronize time plugin, but right now it doesn't function automatically.

I'm not sure why the device would be off by a noticeable amount in any less than a couple of weeks though. That's very odd.

**Comment by NiLuJe** [comment]
I'd double-check if the rtc manages to keep time properly or not (... there is an rtc... right? :D).

e.g., compare `date` to `hwclock -u`

**Comment by WS64** [comment]
@Frenzie, you added "not our bug".
I am not sure.
It is not that the system clock of the Nia is wrong. Nickel got the time right, every time. 
When Koreader show 9:05 in the above example and I go back to Nickel it will tell the correct time, 9:10. When I start Koreader again Koreader too will say 9:10. 
It looks to me as if Koreader does not ask the system clock all the time but only aks that info at start (then it is correct), and after that it calculates its own time with that info. And somehow it loses time doing so. But that's just a guess...

**Comment by Frenzie** [comment]
> It looks to me as if Koreader does not ask the system clock all the time but only aks that info at start (then it is correct), and after that it calculates its own time with that info. And somehow it loses time doing so. But that's just a guess...

KOReader uses os.time() and friends; I have no idea why anyone would do something so silly as that. ;-)

**Comment by WS64** [comment]
@Frenzie, no Wifi is off. I noticed that the time stays right when it is on.
And it loses time in minutes. One hour reading can easily lose 15 or more minutes.

**Comment by Frenzie** [comment]
If you run `date` in the terminal emulator, does it agree with KOReader?

**Comment by NiLuJe** [comment]
Nickel keeps an ntp sync ticking ;).

**Comment by Frenzie** [comment]
> Nickel keeps an ntp sync ticking, so the accuracy of the system clock is irrelevant to it ;).

What do you mean by ticking here? Once a minute or so?

**Comment by NiLuJe** [comment]
I meant that in a vague "periodically" sense, since I don't know the details. On startup and Wi-Fi toggle, for sure.

**Comment by Frenzie** [comment]
Yes, definitely every time wifi connects.

**Comment by Frenzie** [comment]
Btw, I think Settings → Statusbar → Settings → Auto refresh time is disabled by default, in which case the time will only update on page turns.

**Comment by WS64** [comment]
> 
> 
> If you run `date` in the terminal emulator, does it agree with KOReader?

I'll check when it is "off" again!

**Comment by WS64** [comment]
> 
> 
> Btw, I think Settings → Statusbar → Settings → Auto refresh time is disabled by default, in which case the time will only update on page turns. Once a minute or so if enabled.

Yes, I am aware of that, so I alwas check the time right after a page turn

**Comment by NiLuJe** [comment]
> > If you run `date` in the terminal emulator, does it agree with KOReader?
> 
> I'll check when it is "off" again!

When you do:

> I'd double-check if the rtc manages to keep time properly or not (... there is an rtc... right? :D).
> 
> e.g., compare `date` to `hwclock -u`

**Comment by Frenzie** [comment]
> I'll check when it is "off" again!

Ftr, I'd be very surprised if it doesn't. But please also compared it to `hwclock -u`.

**Comment by WS64** [comment]
Actually... How can I check that? I need Wifi for that, and then the time will be right automatically....

**Comment by NiLuJe** [comment]
> If you run `date` in the terminal emulator, does it agree with KOReader?

Via the Terminal plugin.

**Comment by Frenzie** [comment]
You can also use USBNet, but indeed I implicitly meant Tools → More tools → Terminal emulator.

**Comment by WS64** [comment]
17:30 startet reading
17:57: clock shows 17:43
Terminal shows 17:45 (took me 2 minutes to find the terminal)

**Comment by Frenzie** [comment]
So definitely not our bug then. ;-)

**Comment by WS64** [comment]
hwclock shows the correct time

**Comment by WS64** [comment]
About 20 minutes difference bewtween the 2 times

**Comment by Frenzie** [comment]
Then it's some kind of bug in the device's Linux kernel… and Nickel resyncs it as a workaround? That is, `hwclock -su` or something equivalent. I can't tell the difference on any of my devices of course, but on yours it should put the output from `date` and in KOReader back to the correct time then.

**Comment by WS64** [comment]
I guess it is hwclock -s
But, when I do this it stays on "Executing", nothing happens, I cancelled it after 5 minutes.
Tried this twice, first time the Nia was in a strange state. When I shut it down the light would not go off.
I could not execute anything anmore in terminal, every command was immediately followed by "Execution cancelled" (or something like this).
A "reboot device" did show the "rebooting", but then the Nia hung, had to long press the power button.
I have seen this state before a few times, cannot rreally reproduce when it happens. 

If I do it from putty it works, but since I started putty the time was correct again.

If you can tell me where the "page turn" is in the source code I can add the "hwclock -s" there and see if it solves the issue for me. In that case maybe a "if device=Nia then hwclock -s" might be a workaround.

**Comment by Frenzie** [comment]
If you don't add the `-u` I think it'll put your clock to 18:45 (UTC) instead of 19:45 (CET, or replace with another timezone if appropriate).

You can normally cancel the command by tapping outside of the notification btw.

> If you can tell me where the "page turn" is in the source code I can add the "hwclock -s" there and see if it solves the issue for me. In that case maybe a "if device=Nia then hwclock -s" might be a workaround.

A little plugin would be easiest, see for example:

https://github.com/koreader/koreader/blob/c98dfef7fca8b778bc0e43b6bf4c7043892a81a0/plugins/statistics.koplugin/main.lua#L2125

**Comment by WS64** [comment]
os.execute("hwclock -s -u")
does it.
(I had the stats off and read until the clock was 3 minutes back, then I started the stats (with this line added) and the clock was in time again immediately.)

So test successfull.
device.model is "Kobo_luna", is that enough info?

Who will implement it on the corrrect position? Statistics was good for testing, but I have it usually turned off, so it should go to the place that really updates the visible clock in the status bar.

**Comment by Frenzie** [comment]
Is this the latest Kobo firmware version btw? I'm still mightily confused by what's going on here and rather curious what Nickel's actually doing. (An strace might be useful?)

Link to kernel sources for potential reference.
https://github.com/kobolabs/Kobo-Reader/tree/177f41b09f3225e76259063b76b3cf8a3b192c92/hw/imx6ull-nia

**Comment by WS64** [comment]
Actually it is not. I use 4.22.15268, We're up to 4.25.something by now.
Funny enough I forgot to turn off Wifi earlier and probably had some "update automatically" settings on because Kobo did an update to 4.25 this evening. Which left the device unusable, the Kobo menu only returned an error msg, usb mode did not start, nor did koreader. Luckily I had the reader open anyway since I needed to play around with the SD card these last 2 days so often that I could save the Nia by reloading the old firmware again directly to the SD card and then adding Koreader again.
But that also means I am not planning on updating the firmware at all!

I also know that the other guy on the German forum (not sure if it is the same one who reported the error here, the name at least is different) also uses 4.22.15268.

**Comment by WS64** [comment]
Oh, and I don't know what strace means or how to do it, but if you explain I can try!

**Comment by NiLuJe** [comment]
Sidebar: not updating the FW on a recent device is a very good recipe for hitting weird kernel issues like those ;).

**Comment by WS64** [comment]
Yeah, but updating may result in unusable devices! If I hadn't the NIA open anyway I had it returned now since I would not be able to add books to it.
Btw, I actually thought I had the actual FM since when I bought it 2 days ago the first thing it did was doing an update. At least I thought so... Just when I checked the Kobo side after this incident I saw that there are newer FMs.

**Comment by NiLuJe** [comment]
That's because KSM is broken as all hell on Mk. 7 to begin was, and was definitely killed by FW 4.25 ;).

**Comment by Frenzie** [comment]
> Actually it is not. I use 4.22.15268, We're up to 4.25.something by now.

I'd like to know what the behavior is like on 4.25.15875 or higher before committing to a device-specific workaround. :-/ That 4.22.15268 firmware is from when the device only just came out after all.

> Who will implement it on the corrrect position? Statistics was good for testing, but I have it usually turned off, so it should go to the place that really updates the visible clock in the status bar.

I roughly meant that if you copy the Hello World plugin and make it execute that command onPageUpdate, you have a lightweight device-specific workaround that sticks around when updating while we figure out the proper course of action.

**Comment by WS64** [comment]
I am not sure, but does the firmware rely completely on the SD card? If I install the new firmware on a backup card, test, and put the old card bck in will I be back on the old firmware after reboot?

**Comment by Frenzie** [comment]
That should be the case unless it's changed, yes. Iirc there's something like the boot sector, a partition with the Linux kernel/busybox/Nickel, a recovery partition, and a FAT32 partition.

**Comment by WS64** [comment]
4.25.15875 shows the same behavior: clock is too slow, and hwclock -s- -u fixes it.
(And changing the SD card also brought me back the 4.22 firmware, so I have both FMs now. 4.25 looks strange to me though, menus way different than 4.25)

**Comment by Frenzie** [comment]
I dunno, something in this general direction perhaps? (NB That's pseudocode, although it might conceivably also be functional.)

```diff
diff --git a/frontend/device/kobo/device.lua b/frontend/device/kobo/device.lua
index 9eb0c383..20a2dd1d 100644
--- a/frontend/device/kobo/device.lua
+++ b/frontend/device/kobo/device.lua
@@ -269,6 +269,7 @@ local KoboLuna = Kobo:new{
     hasFrontlight = yes,
     touch_snow_protocol = true,
     display_dpi = 212,
+    syncHWTime = true,
 }
 
 function Kobo:init()
@@ -361,6 +362,14 @@ function Kobo:init()
     -- * Turn it off on startup
     -- I've chosen the latter, as I find it vaguely saner, more useful, and it matches Nickel's behavior (I think).
     self:toggleChargingLED(false)
+
+    if self.syncHWTime() then
+        self.syncHWTimeFunc = function()
+            os.execute("hwclock -su")
+            local UIManager = require("ui/uimanager")
+            UIManager:scheduleIn(60, self.syncHWTimeFunc)
+        end
+    end
 end
 
 function Kobo:setDateTime(year, month, day, hour, min, sec)

```

**Comment by WS64** [comment]
Is -su the same as -s -u? I only tested the latter.
And I only executed the hwclock (line 2134 in main.lua of statistics):
```
os.execute("hwclock -s -u")
    local now_ts = os.time()
```
Not sure what the uimanager part does there.
Will this init always (and only) be called when a page is turned or anything else clock relevant is done?

**Comment by Frenzie** [comment]
> Is -su the same as -s -u? I only tested the latter.

Normally those should be equivalent, yes, unless busybox treats 'em differently.

> Will this init always (and only) be called when a page is turned or anything else clock relevant is done?

No, I haven't made up my mind on what the proper course of action would be (e.g., core or plugin, on a timer or on a particular event, using `hwclock` or some `ioctl` I don't know about atm…). This is just some brainstorming.

My concern isn't what works, but what's maintainable. ;-)

**Comment by WS64** [comment]
I think because of the whole statistics a timer base action (once per minute) should be best. Or does that take up too much energy?

**Comment by Frenzie** [comment]
That's why I used `scheduleIn(60, …`. ^_^

**Comment by NiLuJe** [comment]
@Frenzie: FWIW, doing it via FFI is feasible, and you might be able to repurpose some of the rtc wakeup stuff.

c.f., [KoboUSBMS](https://github.com/koreader/KoboUSBMS/blob/143d718fe9161394ded80c77c8749565a402e55d/usbms.c#L1492-L1545), which does the exact inverse of this ;p.

**Comment by Frenzie** [comment]
@NiLuJe I figured as much, but if the performance of os.executing `hwclock` is fine I'd say why bother. Anyway, something along the lines of what I wrote in https://github.com/koreader/koreader/issues/7092#issuecomment-753312105 is roughly what I think would be clearest to read a few years down the line.

**Comment by poire-z** [comment]
> but if the performance of os.executing hwclock is fine I'd say why bother

There's always more risk of failing (because of not-enough memory) with os.execute() and the fork() it does, see https://github.com/koreader/koreader/issues/2239#issuecomment-272188927. So, if it's possible, best to do it via other ways.

In the [above snippet](https://github.com/koreader/koreader/issues/7092#issuecomment-753312105), you need to launch it once explicitely with `self.syncHWTimeFunc()` for it to relaunch itself every 60s :)

**Comment by NiLuJe** [comment]
Well, you admittedly have way more trouble than a drifting clock if a fork OOM's, so, eh :/.

I agree on principle, but since this a dirty hack for a broken kernel/SoC, ngggh ^^.

**Comment by Frenzie** [comment]
> There's always more risk of failing (because of not-enough memory) with os.execute() and the fork() it does, see #2239 (comment). So, if it's possible, best to do it via other ways.

Paraphrased, I almost certainly won't bother experimenting with that. PRs welcome, etc. ;-)

Besides which, as far as issues resulting from OOM goes, having some clock drift is a total non-issue. :-P

**Comment by WS64** [comment]
Since it is theee weeks since the last update, will this one here be fixed with one of the next updates?
Just asking since I have to patch KOreader now after every update to add the os.execute("hwclock -s -u").
It did not cause any issue on my Kia so far, works perfect.

**Comment by NiLuJe** [comment]
"not our bug" ^^.

I'd wait for 4.26 and see if something shakes out, at which point a device-specific crap workaround like what @Frenzie https://github.com/koreader/koreader/issues/7092#issuecomment-753312105 proposed seems the least terrible approach for that.

**Comment by WS64** [comment]
I don't have any hope in 4.26
The bug exists at least since 4.22 and is not fixed in 4.25, why should it be fixed in 4.26, especially since it seems not to affect Nickel?

Even if it is "not our bug" we can easily fix it with a device specific workaround suggested by @Frenzie. Which should not affect other devices, and works already fine for the last 3 weeks on my Nia (although I right now only set the time at page turns).

It does not sound like much, just a clock going wrong, but a wrong clock is really irritating, and of course the whole statistics show wrong results.

**Comment by Frenzie** [comment]
What I wrote can't quite work (although I said general direction of course) due to uimanager/device circular dependency.

**Comment by NiLuJe** [comment]
TL;DR: Patches welcome ;).

**Comment by pazos** [comment]
or report upstream, to @gtalusan

**Comment by WS64** [comment]
> 
> 
> "not our bug" ^^.
> 
> I'd wait for 4.26 and see if something shakes out, at which point a device-specific crap workaround like what @Frenzie [#7092 (comment)](https://github.com/koreader/koreader/issues/7092#issuecomment-753312105) proposed seems the least terrible approach for that.

4.26 does not solve it.

**Comment by NiLuJe** [comment]
I'd ideally like someone else to confirm the issue before any action is taken, because your device is admitedly heavily modified.

But, alas, it's a Nia. And it's hilariously unpopular (for, err, understandable reasons? It's not really appealing compared to a Clara that's barely any more expensive, and it's much *too* expensive compared to an SO Kindle).

Aaaaanyway. I'd probably settle for a dumb workaround, àla spawn a script during the startup process...

**Comment by gtalusan** [comment]
Come on you strace warriors.  It's an easy fix.  Watch what happens approximately 500ms after a wakeup from suspend.

**Comment by NiLuJe** [comment]
Or, that, get me a strace ;o).

(Thanks, George ;p).

**Comment by WS64** [comment]
I have confirmation in a German forum at https://www.e-reader-forum.de/t/neue-firmware-4-26-16704.156628/

**Comment by NiLuJe** [comment]
> Or, that, get me a strace ;o).

1. Install [my bundle o' tools](https://www.mobileread.com/forums/showthread.php?t=254214) to get a shiny strace.

2. Patch our nickel startup script to strace it.
 
```diff

diff --git a/platform/kobo/nickel.sh b/platform/kobo/nickel.sh
index 7cdf9345..aed45b8d 100755
--- a/platform/kobo/nickel.sh
+++ b/platform/kobo/nickel.sh
@@ -91,11 +91,16 @@ if [ -e "/dev/mmcblk1p1" ]; then
     umount /mnt/sd
 fi
 
+# Create a tmpfs for strace's output
+mkdir -p /mnt/log
+umount /mnt/log
+mount -t tmpfs tmpfs /mnt/log -o defaults,size=128M,mode=1777,noatime
+
 # And finally, simply restart nickel.
 # We don't care about horribly legacy stuff, because if people switch between nickel and KOReader in the first place, I assume they're using a decently recent enough FW version.
 # Last tested on an H2O & a Forma running FW 4.7.x - 4.25.x
 /usr/local/Kobo/hindenburg &
-LIBC_FATAL_STDERR_=1 /usr/local/Kobo/nickel -platform kobo -skipFontLoad &
+env LIBC_FATAL_STDERR_=1 strace -fittvTy -o /mnt/log/nickel.log /usr/local/Kobo/nickel -platform kobo -skipFontLoad &
 [ "${PLATFORM}" != "freescale" ] && udevadm trigger &
 
 return 0
```

3. Add a NickelMenu entry that triggers a suspend to save us some time.

```
menu_item : main : Sleep       : power          : sleep
```

4. Start Nickel from KOReader (a.k.a., "Exit"). It'll take longer than normal (because strace ;p).
5. Once it's alive, send it to sleep via your shiny new NM entry.
6. Wait a few minutes.
7. Wake it up.
8. Wait a few seconds more.
9. Start KOReader.

(10. Murder strace in a shell, e.g., `pkill -KILL strace`).

11. Pull the log via SSH (`/mnt/log/nickel.log`).
12. Send it to me :).

**Comment by NiLuJe** [comment]
Now, extra questions because of what I'm seeing on my H2O (which is essentially a systohc on wakeup ^^):

On a cold boot, does it loses time before any kind of suspend kicks in?

**Comment by gtalusan** [comment]
The behaviour is a decade old dating back to the original Kobo with blue button.  The fact that it works whether we need it or not, and we don't really care about it per-device, is a testament to how amazing it is 🤣

**Comment by NiLuJe** [comment]
So, hctosys on wakeup it is then, thanks for the pointers :).

@WS64: That means you're spared a debugging session ^^.
